### ISSUE TITLE
Small optimisations

### ISSUE THREAD

**Issue by purzelrakete** [feldberlin/wavenet#21]
## What

Follow the pytorch tuning guidelines from [here](https://youtu.be/9mS1fIYj1So), as recommended by NVIDIA. Plus some others.

## Why

Will yield a small improvement in training times. Super painless, so why not

## Acceptance Criteria

- [x] Enable cudnn autotuner
- [x] Use efficient gradient zeroing
- [x] Look for pointwise ops to fuse with pytorch JIT

**Comment by purzelrakete** [comment]
## Experiments

[Baseline experiment notebook](https://nbviewer.jupyter.org/github/feldberlin/wavenet/blob/batchnorm/notebooks/experiments/maestro-1630514116.ipynb).

[Treatment experiment notebook](https://nbviewer.jupyter.org/github/feldberlin/wavenet/blob/batchnorm/notebooks/experiments/maestro-1631133331.ipynb). Launched with:

```bash
bin/train maestro -p batch_size 12 -p batch_norm True -p learning_rate 0.05 -p max_epochs 12
```

**Comment by purzelrakete** [comment]
## Results

The metrics at the last step:

|                   | train nll     | test nll |
|-----------|-----------|----------|
| treatment | 1.052        | 1.322       |
| baseline    | 1.575        | 1.453       |

We can say that

- Batch norm overcame the poor init
- Batch norm does not appear to have better training times in this setting, with 2 epochs
- Batch norm does not appear to have better test loss values in this setting, with 2 epochs
- Batch norm is a viable method in this Wavenet, and results are generated largely as before

## Insights

Reducing `n_chans` to 128 and `n_chans_res` to 128 or 96 reduces the number of parameters by a factor of ~4x, while improving performance on a 2 epoch run, resulting in around 3MM parameters. The full model before was setting all channels to 256, which results in a model with around 13MM parameters.

Running for 2 epochs gives results that train something like 10x faster per epoch. The resulting test loss for this network is also better than before. These settings introduce a bottleneck into the residual blocks:

```
(0): ResBlock(
    (conv): Conv1d(128, 192, kernel_size=(2,), stride=(1,))
    (res1x1): Conv1d(96, 128, kernel_size=(1,), stride=(1,))
    (skip1x1): Conv1d(96, 256, kernel_size=(1,), stride=(1,))
  )
```

This suggests that it might be a good idea to look more into this bottleneck. 

Some more insights:

- Batch norm allows a much higher learning rate to be set
- Batch norm removes the original exploding loss problem
- Running experiments with 2 epochs might be a bit misleading. I've switched to 10 on 1 hear of maestro data

## Open Questions

- Can we still run to near zero loss on the track dataset and fully memorise the track? I was using 4x as many parameters before.
